surfactant/infoextractors/docker_tarball_file.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2024 Lawrence Livermore National Security, LLC
+# see: ${repository}/LICENSE
+#
+# SPDX-License-Identifier: MIT
+
+import json
+import tarfile
+from pathlib import PurePosixPath
+from typing import IO, Any, Dict, List, Union
+
+import surfactant.plugin
+from surfactant.sbomtypes import SBOM, Software
+
+
+class optics:
+    class tarball:
+        @staticmethod
+        def manifest_file(tarball: tarfile.TarFile) -> Union[IO[bytes], None]:
+            return tarball.extractfile(
+                {tarinfo.name: tarinfo for tarinfo in tarball.getmembers()}["manifest.json"]
+            )
+
+        @staticmethod
+        def config_file(tarball: tarfile.TarFile, path: str) -> Union[IO[bytes], None]:
+            return tarball.extractfile(
+                {tarinfo.name: tarinfo for tarinfo in tarball.getmembers()}[path]
+            )
+
+    class manifest:
+        @staticmethod
+        def config_path(manifest: List[Dict[str, Any]]) -> List[str]:
+            path = "Config"
+            return [entry[path] for entry in manifest]
+
+        @staticmethod
+        def repo_tags(manifest: List[Dict[str, Any]]) -> List[str]:
+            path = "RepoTags"
+            return [entry[path] for entry in manifest]
+
+
+def portable_path_list(*paths: str):
+    """Convert paths to a portable format acknowledged by"""
+    return tuple([str(PurePosixPath(path_str)) for path_str in paths])
+
+
+def supports_file(filename: str, filetype: str) -> bool:
+    EXPECTED_FILETYPE = "TAR"
+
+    expected_members = portable_path_list(
+        "index.json",
+        "manifest.json",
+        "oci-layout",
+        "repositories",
+        "blobs/sha256",
+    )
```

Comment:
I think this is combining the set of files that are present in the two different container specs (Docker and OCI), which in my limited testing are mutually exclusive so neither type of container will match a check for all of the files being present.

The PR adding Docker Scout support (#193) got merged, which added a check to the [file magic filetypeid code](https://github.com/LLNL/Surfactant/blob/main/surfactant/filetypeid/id_magic.py#L109-L110) to see if a TAR file matches the Docker spec (e.g. has a manifest.json file within it). I think for the Docker config support you're adding we should use the file type id happening there, which will only run once on a tar file (instead of once per plugin).

That said -- I think a similar `is_oci_archive` function should be added to the info extractor that detects if `oci-layout` and `index.json` files, and a `blobs` directory (the spec says it may be empty) are present, and then the id_magic filetypeid code can return a type of `OCI_TAR`.

Reply:
To my knowledge, Docker was based on the OCI specs and OCI they are not mutually exclusive; rather, they are designed to be interoperable, with Docker being one of the primary implementations of OCI-compliant containers. 

If you have any containers that shows them to be mutually exclusive, may I have them for testing please?

Reply:
Mutually exclusive might not be the right term, since there could be a polyglot container image that matches both specs; the set of files present for an archive following the OCI specification (Docker spec v2?) don't also need to be present alongside the files for the Docker (v1.x?) specification. As I was saving images in podman and Docker for testing, I found that with podman there is no overlap in the files output for the two archive types, but [Docker includes manifest.json and repositories files from the old v1.x spec in its OCI archives for backwards compatibility with older Docker versions].

Note: all of these archives are gzipped so that they could be uploaded to GitHub.

podman save --format=oci-archive -o alpine-podman-oci-archive.tar alpine:latest
- [alpine-podman-oci-archive.tar.gz](https://github.com/user-attachments/files/16678141/alpine-podman-oci-archive.tar.gz)

podman save --format=docker-archive -o alpine-podman-docker-archive.tar alpine:latest
- [alpine-podman-docker-archive.tar.gz](https://github.com/user-attachments/files/16678140/alpine-podman-docker-archive.tar.gz)

sudo docker save -o alpine-docker-save.tar alpine:latest `in Docker version <=24.x this matches docker-archive format from podman; 
- [alpine-docker-v24-save.tar.gz](https://github.com/user-attachments/files/16678589/alpine-docker-v24-save.tar.gz)

sudo docker save -o alpine-docker-save.tar alpine:latest `in Docker version >=25.x this is similar to oci-archive format from podman, except with manifest.json and repositories files for backward compatibility)
- [alpine-docker-v27-save.tar.gz](https://github.com/user-attachments/files/16678601/alpine-docker-v27-save.tar.gz)

Reply:
Any thoughts on handling the overlap? The 3 diff cases of Docker && OCI, only Docker and only OCI ?

Reply:
For file type identification, I think OCI can take priority if it is both a valid (older) Docker image and OCI image. Same for information extraction, most of the information should be the same.

I don't have any good ideas for how to handle polyglot files in general -- right now our file type identification plugins assume a file is only one type. To support polyglots we'd probably need to make the file type recognizers start returning lists of types matched.